beluga/include/beluga/random/uniform_grid_distribution.hpp

```diff
@@ -0,0 +1,81 @@
+// Copyright 2024 Ekumen, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef BELUGA_RANDOM_UNIFORM_GRID_DISTRIBUTION_HPP
+#define BELUGA_RANDOM_UNIFORM_GRID_DISTRIBUTION_HPP
+
+#include <random>
+
+#include <range/v3/utility/random.hpp>
+#include <sophus/se2.hpp>
+
+#include <beluga/sensor/data/occupancy_grid.hpp>
+
+namespace beluga {
+
+/// Primary template for a uniform grid distribution.
+/**
+ * \tparam T The result type for sampling from the distribution.
+ */
+template <class T>
+class UniformGridDistribution;
+
+/// Specialization of uniform grid distribution for Sophus::SE2d.
+template <>
+class UniformGridDistribution<Sophus::SE2d> {
+ public:
+  /// Constructs a uniform grid distribution based on the provided occupancy grid.
+  /**
+   * \tparam Derived A type derived from BaseOccupancyGrid2.
+   * \param grid The occupancy grid from which free states will be computed.
+   */
+  template <class Derived>
+  constexpr explicit UniformGridDistribution(const BaseOccupancyGrid2<Derived>& grid)
+      : free_states_{compute_free_states(static_cast<const Derived&>(grid))},
```

Comment:
@nahueespinosa meta: this makes it not uniform, strictly speaking. Don't have a better name for the distribution though :)

Reply:
Yeah, FreeSpaceGridDistribution distribution maybe?

Reply:
Done in https://github.com/Ekumen-OS/beluga/pull/319/commits/f308c57f81651e46182ce99d45bcf0b9b0eafa24!